### PR TITLE
Keyboard Shortcuts: MK1

### PR DESCRIPTION
**Pull request by eshurakov**
### Details
In this PR we're making first steps to support keyboard navigation and improve shortcuts.
* Updated Toggle focus mode shortcut `Cmd + Option + F`
* Updated Focus search field shortcut `Cmd + Shift + F`
* Updated Insert checklist shortcut `Cmd + Shift + C`
* Added Toggle editing content and tags shortcut `Cmd + Shift + Y`
* Added Toggle markdown preview shortcut `Cmd + Shift + P`
* Cycle visible panels shortcut (`Cmd + T`) changed to only toggle sidebar. Cycle panel button on top of the editor stays unchanged and keeps cycling visible panels.
* Arrow navigation between panels (and RLT support)
* Scroll and cursor position is persisted and restored next time the note is opened. It's a best effort and not always precise in case there are changes of note content on another device.

Things that still need to be done:
* Cleanup metadata cache (on logout or when note is deleted)
* Arrow navigation skip hidden panels
* Update selection color based on currently selected panel
* Mystery scroll position issue when using keyboard
* Support navigation using Tab
* Support up / down arrows when in search bar
* Tracks

Ref: #816

### Test
- [x] Check shortcuts from the description
- [x] Toggle markdown preview shortcut is active only when opened note has markdown enabled
- [x] Check arrow navigation (and RTL)
- [x] Check that scroll and cursor position is persisted for a note and restored next time the note is opened

### Review
Only one developer is required to review these changes, but anyone can perform the review.

### CODE REVIEW

File: Simplenote/NoteEditorMetadata.swift [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import Foundation
+
+// MARK: - NoteEditorMetadata
+//
+struct NoteEditorMetadata: Codable {
+    var scrollPosition: CGFloat?
+    var cursorLocation: Int?
```
Comment (jleandroperez) [inline Simplenote/NoteEditorMetadata.swift:7]: Should these be `let` instead?

File: Simplenote/FileStorage.swift [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import Foundation
+
+// MARK: - FileStorage
+//
+class FileStorage<T: Codable> {
```
Comment (jleandroperez) [inline Simplenote/FileStorage.swift:5]: Fancy!!

File: Simplenote/NoteEditorViewController.m [hunk 3]
```diff
@@ -285,42 +272,6 @@ - (void)displayModeWasUpdated:(NSNotification *)notification
...
-    }
-    
-    // Cases 2 and 4
-    if (!beforeCursorMatches && afterCursorMatches) {
-        newCursorLocation += deltaLength;
-    }
-    
-    // Cases 1, 3 and 5 have no change
-    return newCursorLocation;
-}
```
Comment (jleandroperez) [inline Simplenote/NoteEditorViewController.m:322]: Thank you for encapsulating this!!

File: Simplenote/NoteEditorViewController+Swift.swift [hunk 4]
```diff
@@ -831,6 +847,67 @@ extension NoteEditorViewController {
...
+        metadataCache.store(cursorLocation: noteEditor.selectedRange().location, for: simperiumKey)
+    }
+
+    @objc
+    func restoreCursorLocation() {
+        guard let simperiumKey = note?.simperiumKey,
+              let cursorLocation = metadataCache.metadata(for: simperiumKey)?.cursorLocation else {
+            return
+        }
+        noteEditor.setSelectedRange(NSRange(location: cursorLocation, length: 0))
```
Comment (jleandroperez) [inline Simplenote/NoteEditorViewController+Swift.swift:906]: Just noting that (I've just verified!!) that if the Cursor Location isn't really valid, NSTextView doesn't crash.

(I was thinking: what if there are unaccounted scenarios that could render invalid the old Location)

File: Simplenote/NoteEditorViewController.m [hunk 5]
```diff
@@ -224,12 +215,8 @@ - (void)displayNote:(Note *)selectedNote
...
     [self.storage refreshStyleWithMarkdownEnabled:self.note.markdown];
 
-    NSValue *lastKnownScrollOffset = self.noteScrollPositions[selectedNote.simperiumKey];
-    if (lastKnownScrollOffset != nil) {
-        [self.scrollView.documentView scrollPoint:lastKnownScrollOffset.pointValue];
-    } else {
-        [self.scrollView scrollToTopWithAnimation:NO];
-    }
+    [self restoreScrollPosition];
+    [self restoreCursorLocation];
```
Comment (jleandroperez) [inline Simplenote/NoteEditorViewController.m:219]: Looks so much better!!

**Review by jleandroperez — APPROVED** [review]
**SUPER solid** (as usual!!!)

:shipit:

File: Simplenote/NoteEditorMetadata.swift (hunk 1 above)
Comment (eshurakov) [reply Simplenote/NoteEditorMetadata.swift:7]: Made it `let` and also combined save cursor and scroll position calls into one 👍 
https://github.com/Automattic/simplenote-macos/pull/888/commits/b1025f0c34eb3e2aec597192ba3db047345ed25f

**Comment by pachlava** [comment]
[testing] fix/feature verified on MBP 16" Intel | macOS Big Sur 11.2 | Simplenote 2.9 (11223)

**Pull request merged**